### ISSUE TITLE
Depending on abstractions rather than implementations

### ISSUE THREAD

**Issue by jbieliauskas** [digitronas/akeneo-coding-standard#17]
Let's discuss the "D" in "SOLID". 

It's true that it is better to accept interfaces, not concrete classes to our functions. Using a concrete class inevitably means assuming something about the implementation, which breaks encapsulation. Also, when using interfaces, it's easier to extend functionality, because instead of changing the code in-place, we simply substitute the concrete object with another one.

There are exceptions - some object are simply used for convenience, to facilitate some logic. But even those objects are subject to change. For instance, we could decorate the *convenient* object and add logging functionality to it. Or we could decide not to use convenience objects at all.

Bottom line is, I think we should ignore exceptions like these and make a rule that says "All classes should implement an interfaces".
